File: 208-ImplementTriePrefixTree/208-ImplementTriePrefixTree.js
// Last updated: 4/29/2026, 10:32:56 AM

var Trie = function() {
    this.dic = {};
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let curr = this.dic;
    for(const ch of word) {
        if(curr[ch] === undefined) {
            curr[ch] = {};
        }
        curr = curr[ch];
    }
    curr['wordEnd'] = true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    let curr = this.dic;
    for(const ch of word) {
        if(curr[ch] === undefined) {
            return false;
        }
        curr = curr[ch];
    }
    return curr['wordEnd'] === true;
};

/** 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    let curr = this.dic;
    for(const ch of prefix) {
        if(curr[ch] === undefined) {
            return false;
        }
        curr = curr[ch];
    }
    return true;
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: 208-ImplementTriePrefixTree/208-ImplementTriePrefixTree.py
# Last updated: 4/29/2026, 10:33:01 AM
class Trie:

    def __init__(self):
        self.root = {}

    def insert(self, word: str) -> None:
        # print(word)
        node = self.root
        for ch in word:
            if node.get(ch) == None:
                node[ch] = {}
            node = node[ch]
        node['complete'] = True
        # print(self.root)

    def search(self, word: str) -> bool:
        node = self.root
        for ch in word:
            if node.get(ch) == None:
                return False;
            node = node[ch]
        if node.get('complete') == True:
            return True
        else: 
            return False

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for ch in prefix:
            if node.get(ch) == None:
                return False;
            node = node[ch]
        return True
        
# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)